fusillade/clouddirectory.py

```diff
@@ -0,0 +1,807 @@
+                                                  MaxResults=self._page_limit)
+            else:
+                break
+
+    def create_object(self, link_name: str, statement: str, obj_type: str, **kwargs) -> str:
+        operations = list()
+        object_attribute_list = self._get_object_attribute_list(facet=obj_type, **kwargs)
+        parent_path = self.get_obj_type_path(obj_type)
+        operations.append(self.batch_create_object(parent_path, link_name, obj_type, object_attribute_list))
+        object_ref = parent_path + link_name
+
+        object_attribute_list = self.get_policy_attribute_list(f'{obj_type}Policy', statement, Statement=statement)
+        policy_link_name = f"{link_name}_policy"
+        parent_path = self.get_obj_type_path('policy')
+        operations.append(self.batch_create_object(parent_path,
+                                                   policy_link_name,
+                                                   'IAMPolicy',
+                                                   object_attribute_list))
+        policy_ref = parent_path + policy_link_name
+
+        operations.append(self.batch_attach_policy(policy_ref, object_ref))
+        self.batch_write(operations)  # TODO check that everything passed
+
+        return object_ref, policy_ref
+
+    def get_object_attrtibutes(self, obj_ref: str, facet: str, attributes: typing.List[str]) -> typing.Dict[str, str]:
+        return ad.get_object_attributes(DirectoryArn=self._dir_arn,
+                                        ObjectReference={'Selector': obj_ref},
+                                        SchemaFacet={
+                                            'SchemaArn': self._schema,
+                                            'FacetName': facet
+                                        },
+                                        AttributeNames=attributes
+                                        )
+
+    def _get_object_attribute_list(self, facet="User", **kwargs) -> typing.List[typing.Dict[str, typing.Any]]:
+        return [dict(Key=dict(SchemaArn=self._schema, FacetName=facet, Name=k), Value=dict(StringValue=v))
+                for k, v in kwargs.items()]
+
+    def get_policy_attribute_list(self,
+                                  policy_type: str,
+                                  policy_document: str,
+                                  facet: str = "IAMPolicy",
+                                  **kwargs) -> typing.List[typing.Dict[str, typing.Any]]:
+        obj = self._get_object_attribute_list(facet=facet, **kwargs)
+        obj.append(dict(Key=dict(
+            SchemaArn=self._schema,
+            FacetName=facet,
+            Name='policy_type'),
+            Value=dict(StringValue=policy_type)))
+        obj.append(
+            dict(Key=dict(SchemaArn=self._schema,
+                          FacetName="IAMPolicy",
+                          Name="policy_document"),
+                 Value=dict(BinaryValue=json.dumps(policy_document).encode())))
```

Comment:
Why is the above being added as a `StringValue` and this one added as a `BinaryValue`?

Reply:
"Policies" are a special type of Node that have a unique scheme that require the fields policy_name and policy_document. This was discovered by trial and error that a BinaryValue is required for policy_document. I couldn't find it in their documentation why this was.

Reply:
That's pretty unusual.  Definitely worth documenting here.

Reply:
adding doc